File: CAPSquadron_API/Models/QualityCadetUnitReport.cs
using System.ComponentModel.DataAnnotations.Schema;

namespace CAPSquadron_API.Models;

public class QualityCadetUnitReport
{
    public int Id { get; set; }
    public string? Charter { get; set; }
    public int CurrentCadets { get; set; }

    [Column("enrollment_25_plus_cadets")]
    public string? Enrollment25PlusCadets { get; set; }

    public int CadetsJoined { get; set; }

    [Column("curry_in_8_weeks")]
    public int CurryIn8Weeks { get; set; }

    [Column("onboarding_70_percent")]
    public string? Onboarding70Percent { get; set; }

    [Column("cadets_with_wb")]
    public int CadetsWithWB { get; set; }

    [Column("cadet_achievement_45_percent")]
    public string? CadetAchievement45Percent { get; set; }

    [Column("cadets_with_oflights")]
    public int CadetsWithOFlights { get; set; }

    [Column("oflights_70_percent")]
    public string? OFlights70Percent { get; set; }

    [Column("cadets_with_encamp")]
    public int CadetsWithEncamp { get; set; }

    [Column("encamp_50_percent")]
    public string? Encamp50Percent { get; set; }

    [Column("cadets_with_ges")]
    public int CadetsWithGES { get; set; }

    [Column("ges_60_percent")]
    public string? GES60Percent { get; set; }

    public string? AEX { get; set; }
    public string? STEM { get; set; }

    [Column("ae_aex_or_stem_kit")]
    public string? AEAexOrStemKit { get; set; }

    [Column("outside_activities")]
    public string? OutsideActivities { get; set; }

    [Column("seniors_with_tlc")]
    public int SeniorsWithTLC { get; set; }

    [Column("adult_leadership_3_plus_tlc_grads")]
    public string? AdultLeadership3PlusTlcGrads { get; set; }

    [Column("seniors_with_cp_specialty_track_rating")]
    public int SeniorsWithCPSpecialtyTrackRating { get; set; }

    [Column("specialty_track_2_plus_seniors_with_rating")]
    public string? SpecialtyTrack2PlusSeniorsWithRating { get; set; }

    [Column("qcua_6_plus_criteria_met")]
    public string? QCUA6PlusCriteriaMet { get; set; }

    [Column("number_of_criteria_met")]
    public int NumberOfCriteriaMet { get; set; }

    [Column("report_date")]
    public DateTime ReportDate { get; set; }
}

File: CAPSquadron_ETL/services/quality_cadet_unit_report.py
# services/quality_cadet_unit_report.py

from sqlalchemy.orm import Session
from fastapi import HTTPException, UploadFile
from datetime import datetime
from crud.quality_cadet_unit_report import create_quality_cadet_unit_report
from schemas.quality_cadet_unit_report import QualityCadetUnitReportCreate
from utils.file_processing import process_file

def process_quality_cadet_unit_report(file: UploadFile):
    df = process_file(file)
    return df

def save_quality_cadet_unit_report_to_db(df, db: Session, report_date: datetime):
    for _, row in df.iterrows():
        report_data = QualityCadetUnitReportCreate(
            charter=row['Charter'],
            current_cadets=row['Current Cadets'],
            enrollment_25_plus_cadets=row['Enrollment (25+ Cadets)'],
            cadets_joined=row['FirstCadets Joined 31 Aug 2023 - 31 Aug 2024'],
            curry_in_8_weeks=row['Curry in 8 Wks'],
            onboarding_70_percent=row['Onboarding (70% Curry in 8 Wks)'],
            cadets_with_wb=row['Cadets w/ WB'],
            cadet_achievement_45_percent=row['Cadet Achv.(45% w/ WB)'],
            cadets_with_oflights=row['Cadets w/ O\'Flights'],
            oflights_70_percent=row['O\'Flights (70% w/ First Flight)'],
            cadets_with_encamp=row['Cadets w/Encamp'],
            encamp_50_percent=row['Encamp (50% w/ Encamp)'],
            cadets_with_ges=row['Cadets w/ GES'],
            ges_60_percent=row['GES (60% w/ GES)'],
            aex=row['AEX'],
            stem=row['STEM'],
            ae_aex_or_stem_kit=row['AE (AEX or STEM kit)'],
            outside_activities=row['Outside Activities'],
            seniors_with_tlc=row['Seniors w/ TLC'],
            adult_leadership_3_plus_tlc_grads=row['Adult Leadership (3+ TLC Grads)'],
            seniors_with_cp_specialty_track_rating=row['Seniors w/ CP Specialty Track Rating'],
            specialty_track_2_plus_seniors_with_rating=row['Specialty Track (2+ Seniors w/Rating)'],
            qcua_6_plus_criteria_met=row['QCUA (6+ Criteria Met)'],
            number_of_criteria_met=row['No. of Criteria Met'],
            report_date=report_date
        )
        create_quality_cadet_unit_report(db, report_data)
